File: code_puppy/agents/event_stream_handler.py
"""Event stream handler for processing streaming events from agent runs."""

import asyncio
import importlib.util
import logging
from collections.abc import AsyncIterable
from typing import Any

from pydantic_ai import PartDeltaEvent, PartEndEvent, PartStartEvent, RunContext
from pydantic_ai.messages import (
    TextPart,
    TextPartDelta,
    ThinkingPart,
    ThinkingPartDelta,
    ToolCallPart,
    ToolCallPartDelta,
)
from rich.console import Console
from rich.markup import escape
from rich.text import Text

from code_puppy.agents.stream_event_normalizer import normalize_stream_event
from code_puppy.config import get_banner_color, get_subagent_verbose
from code_puppy.messaging.spinner import pause_all_spinners, resume_all_spinners
from code_puppy.tools.subagent_context import is_subagent

logger = logging.getLogger(__name__)

# Module-level buffer for batching stream events
_pending_stream_events: list[tuple[str, Any, Any]] = []
_STREAM_FLUSH_INTERVAL = 50

# Module-level state for stream-to-render coordination
# Tracks whether text was streamed AND how many terminal lines were printed
_streamed_line_count: int = 0  # Lines printed during streaming (including banner)
_did_stream_text: bool = False  # Whether text content was streamed


def get_stream_state() -> tuple[bool, int]:
    """Return (did_stream, line_count) and reset state.

    This is used to coordinate between the streaming handler and the
    message renderer. The streaming handler tracks how many lines were
    printed, and the renderer uses this to erase them before re-rendering
    as markdown.

    Returns:
        Tuple of (whether text was streamed, number of terminal lines printed)
    """
    global _did_stream_text, _streamed_line_count
    result = (_did_stream_text, _streamed_line_count)
    _did_stream_text = False
    _streamed_line_count = 0
    return result


def _reset_stream_state() -> None:
    """Reset the stream state at the start of a new handler invocation."""
    global _did_stream_text, _streamed_line_count
    _did_stream_text = False
    _streamed_line_count = 0


# Track active flush tasks to prevent garbage collection and ensure completion
_active_flush_tasks: set[asyncio.Task] = set()

# Flush text-part delta buffer when it exceeds this many characters.
# Larger than the TUI's 20-char threshold because terminal ANSI writes
# benefit from batching; TUI updates a widget which has different perf
# characteristics.
_TEXT_FLUSH_CHAR_THRESHOLD = 80


def _fire_stream_event(event_type: str, event_data: Any) -> None:
    """Fire a stream event callback asynchronously (non-blocking) with batching.

    Events are normalized to a unified schema and batched to reduce task
    creation overhead. The batch is flushed when 'part_end' event is received
    or when batch size reaches the threshold.

    Args:
        event_type: Type of the event (e.g., 'part_start', 'part_delta', 'part_end')
        event_data: Data associated with the event (will be normalized)
    """
    global _pending_stream_events

    try:
        # Check if callbacks module is available without importing
        if importlib.util.find_spec("code_puppy.callbacks") is None:
            raise ImportError("callbacks module not available")
        from code_puppy.messaging import get_session_context

        agent_session_id = get_session_context()

        # Normalize event data to unified schema for consistent processing
        # by downstream consumers like Agent Trace
        if isinstance(event_data, dict):
            normalized_data = normalize_stream_event(event_type, event_data)
        else:
            # Fallback for non-dict event data (shouldn't happen in practice)
            normalized_data = {
                "content_delta": str(event_data) if event_data else None,
                "args_delta": None,
                "tool_name": None,
                "tool_name_delta": None,
                "part_kind": "unknown",
                "index": -1,
                "raw": {"_original": event_data},
            }

        _pending_stream_events.append((event_type, normalized_data, agent_session_id))

        # Flush on part_end or when batch threshold reached
        if (
            event_type == "part_end"
            or len(_pending_stream_events) >= _STREAM_FLUSH_INTERVAL
        ):
            batch = _pending_stream_events.copy()
            _pending_stream_events.clear()
            task = asyncio.create_task(_flush_stream_events(batch))
            _active_flush_tasks.add(task)
            task.add_done_callback(_active_flush_tasks.discard)
    except ImportError:
        logger.debug("callbacks or messaging module not available for stream event")
    except Exception as e:
        logger.debug(f"Error firing stream event callback: {e}")


async def _flush_stream_events(batch: list) -> None:
    """Flush a batch of stream events to the callbacks.

    Args:
        batch: List of (event_type, event_data, session_id) tuples to process.
    """
    from code_puppy import callbacks

    for event_type, event_data, session_id in batch:
        try:
            await callbacks.on_stream_event(event_type, event_data, session_id)
        except Exception as e:
            logger.debug(f"Error flushing stream event: {e}")


async def _drain_pending_stream_events() -> None:
    """Drain any pending stream events before handler exits.

    Ensures that any batched events are delivered before the handler completes.
    Also awaits any in-flight flush tasks to prevent event loss on shutdown.
    """
    global _pending_stream_events

    if _pending_stream_events:
        batch = _pending_stream_events.copy()
        _pending_stream_events.clear()
        await _flush_stream_events(batch)

    # Await any in-flight flush tasks
    if _active_flush_tasks:
        await asyncio.gather(*_active_flush_tasks, return_exceptions=True)
        _active_flush_tasks.clear()


# Module-level console for streaming output
# Set via set_streaming_console() to share console with spinner
_streaming_console: Console | None = None


def set_streaming_console(console: Console | None) -> None:
    """Set the console used for streaming output.

    Sharing a console with the spinner keeps spinner pause/resume
    coordination working during thinking-part and tool-call output.
    Text parts stream as escaped plain chunks (no Rich Live wrapper)
    to avoid the cascading re-render bug.

    Args:
        console: The Rich console to use, or None to use a fallback.
    """
    global _streaming_console
    _streaming_console = console


def get_streaming_console() -> Console:
    """Get the console for streaming output.

    Returns the configured console or creates a fallback Console.
    """
    if _streaming_console is not None:
        return _streaming_console
    return Console()


def _should_suppress_output() -> bool:
    """Check if sub-agent output should be suppressed.

    Returns:
        True if we're in a sub-agent context and verbose mode is disabled.
    """
    return is_subagent() and not get_subagent_verbose()


async def event_stream_handler(ctx: RunContext, events: AsyncIterable[Any]) -> None:
    """Handle streaming events from the agent run.

    This function processes streaming events and emits TextPart, ThinkingPart,
    and ToolCallPart content with styled banners/tokens as they stream in.

    Text parts are streamed as escaped plain text (not live markdown) to avoid
    Rich Live re-render duplication that occurs with long content. The cascading
    duplication bug happened because Rich Live with
    vertical_overflow="visible" re-emits the entire buffer on every refresh when
    content exceeds terminal height. Now we use list-buffered delta streaming with
    flush-on-newline/80-chars, similar to the TUI approach in stream_renderer.py.
    Final markdown formatting happens elsewhere via the message bus (AgentResponseMessage).
    ThinkingPart and ToolCallPart streaming behavior is unchanged.

    Args:
        ctx: The run context.
        events: Async iterable of streaming events (PartStartEvent, PartDeltaEvent, etc.).
    """
    global _did_stream_text, _streamed_line_count

    # Reset stream state at the start of each handler invocation
    _reset_stream_state()

    try:
        # If we're in a sub-agent and verbose mode is disabled, silently consume events
        if _should_suppress_output():
            async for _ in events:
                pass  # Just consume events without rendering
            return

        # Use the module-level console (set via set_streaming_console)
        console = get_streaming_console()

        # Track which part indices we're currently streaming (for Text/Thinking/Tool parts)
        streaming_parts: set[int] = set()
        thinking_parts: set[int] = (
            set()
        )  # Track which parts are thinking (for dim style)
        text_parts: set[int] = set()  # Track which parts are text
        tool_parts: set[int] = set()  # Track which parts are tool calls
        banner_printed: set[int] = set()  # Track if banner was already printed
        token_count: dict[int, int] = {}  # Track token count per text/tool part
        tool_names: dict[int, str] = {}  # Track tool name per tool part index
        did_stream_anything = False  # Track if we streamed any content

        # Text part streaming state: list-buffered for O(1) append, flush on newline
        text_buffers: dict[int, list[str]] = {}  # Accumulated content chunks per index

        async def _print_thinking_banner() -> None:
            """Print the THINKING banner with spinner pause and line clear."""
            nonlocal did_stream_anything

            pause_all_spinners()
            await asyncio.sleep(0.1)  # Delay to let spinner fully clear
            # Clear line and print newline before banner
            console.print(" " * 50, end="\r")
            console.print()  # Newline before banner
            # Bold banner with configurable color and lightning bolt
            thinking_color = get_banner_color("thinking")
            console.print(
                Text.from_markup(
                    f"[bold white on {thinking_color}] THINKING [/bold white on {thinking_color}] [dim]\\u26a1 "
                ),
                end="",
            )
            did_stream_anything = True

        async def _print_response_banner() -> None:
            """Print the AGENT RESPONSE banner with spinner pause and line clear."""
            nonlocal did_stream_anything
            global _did_stream_text, _streamed_line_count

            pause_all_spinners()
            await asyncio.sleep(0.1)  # Delay to let spinner fully clear
            # Clear line and print newline before banner
            console.print(" " * 50, end="\r")
            console.print()  # Newline before banner
            response_color = get_banner_color("agent_response")
            console.print(
                Text.from_markup(
                    f"[bold white on {response_color}] AGENT RESPONSE [/bold white on {response_color}]"
                )
            )
            did_stream_anything = True
            # Track that we streamed content and count lines
            # Banner takes 2 lines: the blank line before it + the banner itself
            _did_stream_text = True
            _streamed_line_count += 2

        async for event in events:
            # PartStartEvent - register the part but defer banner until content arrives
            if isinstance(event, PartStartEvent):
                # Fire stream event callback for part_start
                _fire_stream_event(
                    "part_start",
                    {
                        "index": event.index,
                        "part_type": type(event.part).__name__,
                        "part": event.part,
                    },
                )

                part = event.part
                if isinstance(part, ThinkingPart):
                    streaming_parts.add(event.index)
                    thinking_parts.add(event.index)
                    # If there's initial content, print banner + content now
                    if part.content and part.content.strip():
                        await _print_thinking_banner()
                        escaped = escape(part.content)
                        console.print(f"[dim]{escaped}[/dim]", end="")
                        banner_printed.add(event.index)
                elif isinstance(part, TextPart):
                    streaming_parts.add(event.index)
                    text_parts.add(event.index)
                    # Initialize text buffer for this text part (list for O(1) append)
                    text_buffers[event.index] = []
                    # Handle initial content if present
                    if part.content and part.content.strip():
                        await _print_response_banner()
                        banner_printed.add(event.index)
                        # Immediately print initial content as plain text
                        console.print(part.content, end="", markup=False)
                elif isinstance(part, ToolCallPart):
                    streaming_parts.add(event.index)
                    tool_parts.add(event.index)
                    token_count[event.index] = 0  # Initialize token counter
                    # Capture tool name from the start event
                    tool_names[event.index] = part.tool_name or ""
                    # Track tool name for display
                    banner_printed.add(
                        event.index
                    )  # Use banner_printed to track if we've shown tool info

            # PartDeltaEvent - stream the content as it arrives
            elif isinstance(event, PartDeltaEvent):
                # Fire stream event callback for part_delta
                _fire_stream_event(
                    "part_delta",
                    {
                        "index": event.index,
                        "delta_type": type(event.delta).__name__,
                        "delta": event.delta,
                    },
                )

                if event.index in streaming_parts:
                    delta = event.delta
                    if isinstance(delta, (TextPartDelta, ThinkingPartDelta)):
                        if delta.content_delta:
                            # For text parts, stream as plain escaped text (no Live)
                            if event.index in text_parts:
                                # Print banner on first content
                                if event.index not in banner_printed:
                                    await _print_response_banner()
                                    banner_printed.add(event.index)

                                # Append content to list buffer
                                text_buffers[event.index].append(delta.content_delta)
                                # Flush on newline or when buffer exceeds threshold
                                buf = "".join(text_buffers[event.index])
                                if "\n" in buf or len(buf) > _TEXT_FLUSH_CHAR_THRESHOLD:
                                    console.print(buf, end="", markup=False)
                                    # Count newlines in the flushed buffer
                                    _streamed_line_count += buf.count("\n")
                                    text_buffers[event.index] = []
                            else:
                                # For thinking parts, stream immediately (dim)
                                if event.index not in banner_printed:
                                    await _print_thinking_banner()
                                    banner_printed.add(event.index)
                                escaped = escape(delta.content_delta)
                                console.print(f"[dim]{escaped}[/dim]", end="")
                    elif isinstance(delta, ToolCallPartDelta):
                        # For tool calls, estimate tokens from args_delta content
                        # args_delta contains the streaming JSON arguments
                        args_delta = getattr(delta, "args_delta", "") or ""
                        if args_delta:
                            # Rough estimate: 4 chars ≈ 1 token (same heuristic as subagent_stream_handler)
                            estimated_tokens = max(1, len(args_delta) // 4)
                            token_count[event.index] += estimated_tokens
                        else:
                            # Even empty deltas count as activity
                            token_count[event.index] += 1

                        # Update tool name if delta provides more of it
                        tool_name_delta = getattr(delta, "tool_name_delta", "") or ""
                        if tool_name_delta:
                            tool_names[event.index] = (
                                tool_names.get(event.index, "") + tool_name_delta
                            )

                        # Use stored tool name for display
                        tool_name = tool_names.get(event.index, "")
                        count = token_count[event.index]
                        # Display with tool wrench icon and tool name
                        if tool_name:
                            console.print(
                                f"  \U0001f527 Calling {tool_name}... {count} token(s)   ",
                                end="\r",
                            )
                        else:
                            console.print(
                                f"  \U0001f527 Calling tool... {count} token(s)   ",
                                end="\r",
                            )

            # PartEndEvent - finish the streaming with a newline
            elif isinstance(event, PartEndEvent):
                # Fire stream event callback for part_end
                _fire_stream_event(
                    "part_end",
                    {
                        "index": event.index,
                        "next_part_kind": getattr(event, "next_part_kind", None),
                    },
                )

                if event.index in streaming_parts:
                    # For text parts, flush any remaining buffered content and add newline
                    if event.index in text_parts:
                        # Flush any remaining buffered content before cleanup
                        if event.index in text_buffers:
                            remaining = text_buffers.pop(event.index, None)
                            if remaining:
                                buf = "".join(remaining)
                                if buf:
                                    console.print(buf, end="", markup=False)
                                    # Count remaining newlines in the final buffer
                                    _streamed_line_count += buf.count("\n")
                        # Print trailing newline only if banner was printed (i.e., we had content)
                        if event.index in banner_printed:
                            console.print()  # Final newline after text streaming
                            _streamed_line_count += 1  # Count the final newline
                    # For tool parts, clear the chunk counter line
                    elif event.index in tool_parts:
                        # Clear the chunk counter line by printing spaces and returning
                        console.print(" " * 50, end="\r")
                    # For thinking parts, just print newline
                    elif event.index in banner_printed:
                        console.print()  # Final newline after streaming

                    # Clean up token count and tool names
                    token_count.pop(event.index, None)
                    tool_names.pop(event.index, None)
                    # Clean up all tracking sets
                    streaming_parts.discard(event.index)
                    thinking_parts.discard(event.index)
                    text_parts.discard(event.index)
                    tool_parts.discard(event.index)
                    banner_printed.discard(event.index)

                    # Resume spinner if next part is NOT text/thinking/tool (avoid race condition)
                    # If next part is None or handled differently, it's safe to resume
                    # Note: spinner itself handles blank line before appearing
                    next_kind = getattr(event, "next_part_kind", None)
                    if next_kind not in ("text", "thinking", "tool-call"):
                        resume_all_spinners()

    finally:
        # Force cursor visibility restoration (defensive hygiene)
        try:
            console.show_cursor(True)
        except Exception:
            pass

    # Spinner is resumed in PartEndEvent when appropriate (based on next_part_kind)
    # Drain any remaining buffered stream events before handler exits
    await _drain_pending_stream_events()

File: elixir/code_puppy_control/lib/code_puppy_control/agent/event_stream_handler.ex
defmodule CodePuppyControl.Agent.EventStreamHandler do
  @moduledoc """
  UI streaming bridge for the main agent.

  Processes canonical stream events from an agent run, publishing them
  to the EventBus for TUI/WebSocket consumers and firing `stream_event`
  callbacks for plugins.

  ## Design

  This GenServer replaces the Python `event_stream_handler.py`'s
  async function with an Elixir-native process. Key differences:

  - **EventBus replaces Rich console** — Instead of writing to a Rich
    Console, events are published via PubSub so TUI and WebSocket
    subscribers receive them.
  - **Process state replaces module globals** — Stream state (line count,
    banner tracking, text buffers) lives in the GenServer state rather
    than module-level variables.
  - **Batched callback delivery** — Events are batched before firing
    `stream_event` callbacks to reduce process message overhead.

  ## Usage

      # Start a handler for an agent run
      {:ok, pid} = EventStreamHandler.start_link(
        run_id: "run-123",
        session_id: "session-456"
      )

      # Feed canonical stream events
      EventStreamHandler.push(pid, %TextDelta{index: 0, text: "Hello"})
      EventStreamHandler.push(pid, %TextEnd{index: 0, id: nil})

      # Drain remaining events and stop
      EventStreamHandler.drain(pid)

  Port of `code_puppy/agents/event_stream_handler.py`.
  """

  use GenServer

  require Logger

  alias CodePuppyControl.{Callbacks, EventBus}
  alias CodePuppyControl.Stream.{Event, EventNormalizer}

  # Flush the callback batch when it reaches this size
  @stream_flush_interval 50
  # Flush text content when buffer exceeds this many characters
  @text_flush_char_threshold 80

  # ── State ──────────────────────────────────────────────────────────

  @type t :: %__MODULE__{
          run_id: String.t() | nil,
          session_id: String.t() | nil,
          # Part tracking
          streaming_parts: MapSet.t(non_neg_integer()),
          thinking_parts: MapSet.t(non_neg_integer()),
          text_parts: MapSet.t(non_neg_integer()),
          tool_parts: MapSet.t(non_neg_integer()),
          banner_printed: MapSet.t(non_neg_integer()),
          # Text buffering: index → list of chunks
          text_buffers: %{non_neg_integer() => [String.t()]},
          # Tool tracking: index → token count
          token_count: %{non_neg_integer() => non_neg_integer()},
          # Tool tracking: index → tool name
          tool_names: %{non_neg_integer() => String.t()},
          # Stream state for UI coordination
          did_stream_text: boolean(),
          streamed_line_count: non_neg_integer(),
          # Batched callback events
          pending_events: [{String.t(), map(), String.t() | nil}]
        }

  defstruct [
    :run_id,
    :session_id,
    streaming_parts: MapSet.new(),
    thinking_parts: MapSet.new(),
    text_parts: MapSet.new(),
    tool_parts: MapSet.new(),
    banner_printed: MapSet.new(),
    text_buffers: %{},
    token_count: %{},
    tool_names: %{},
    did_stream_text: false,
    streamed_line_count: 0,
    pending_events: []
  ]

  # ── Public API ─────────────────────────────────────────────────────

  @doc """
  Starts the event stream handler.

  ## Options

    * `:run_id` — Run identifier for EventBus routing
    * `:session_id` — Session identifier for EventBus routing
  """
  @spec start_link(keyword()) :: GenServer.on_start()
  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts)
  end

  @doc """
  Pushes a canonical stream event to the handler for processing.
  """
  @spec push(pid(), Event.canonical()) :: :ok
  def push(pid, event) do
    GenServer.cast(pid, {:push, event})
  end

  @doc """
  Returns the current stream state and resets it.

  Returns `{did_stream, line_count}` for coordination with the
  message renderer — the renderer uses line_count to erase
  streamed lines before re-rendering as markdown.
  """
  @spec get_stream_state(pid()) :: {boolean(), non_neg_integer()}
  def get_stream_state(pid) do
    GenServer.call(pid, :get_stream_state)
  end

  @doc """
  Drains any pending batched callback events and stops the handler.
  """
  @spec drain(pid()) :: :ok
  def drain(pid) do
    GenServer.call(pid, :drain)
  end

  @doc """
  Sets the streaming console reference (no-op in Elixir — kept for API
  compatibility; UI rendering happens via EventBus subscribers).
  """
  @spec set_streaming_console(pid(), term()) :: :ok
  def set_streaming_console(_pid, _console), do: :ok

  # ── GenServer Callbacks ─────────────────────────────────────────────

  @impl true
  def init(opts) do
    state = %__MODULE__{
      run_id: Keyword.get(opts, :run_id),
      session_id: Keyword.get(opts, :session_id)
    }

    {:ok, state}
  end

  @impl true
  def handle_cast({:push, event}, state) do
    state = process_event(event, state)
    {:noreply, state}
  end

  @impl true
  def handle_call(:get_stream_state, _from, state) do
    result = {state.did_stream_text, state.streamed_line_count}
    state = %{state | did_stream_text: false, streamed_line_count: 0}
    {:reply, result, state}
  end

  @impl true
  def handle_call(:drain, _from, state) do
    # Flush any remaining batched events
    state = flush_pending_events(state)
    {:stop, :normal, :ok, state}
  end

  # ── Event Processing ────────────────────────────────────────────────

  defp process_event(%Event.TextStart{index: idx} = event, state) do
    fire_callback(
      "part_start",
      %{
        "index" => idx,
        "part_type" => "TextPart",
        "part" => event
      },
      state
    )

    state
    |> add_to_set(:streaming_parts, idx)
    |> add_to_set(:text_parts, idx)
    |> put_text_buffer(idx, [])
    |> maybe_emit_thinking_banner(event, "TextPart")
  end

  defp process_event(%Event.TextDelta{index: idx, text: text} = _event, state) do
    fire_callback(
      "part_delta",
      %{
        "index" => idx,
        "delta_type" => "TextPartDelta",
        "delta" => %{content_delta: text}
      },
      state
    )

    # Only process if this is a tracked text part
    if MapSet.member?(state.text_parts, idx) do
      state
      |> maybe_emit_response_banner(idx)
      |> append_text_buffer(idx, text)
      |> maybe_flush_text_buffer(idx)
    else
      state
    end
  end

  defp process_event(%Event.TextEnd{index: idx} = _event, state) do
    fire_callback(
      "part_end",
      %{
        "index" => idx,
        "next_part_kind" => nil
      },
      state
    )

    # Flush any remaining buffered content
    state = flush_remaining_text_buffer(state, idx)

    # Add trailing newline if we printed a banner (i.e., had content)
    state =
      if MapSet.member?(state.banner_printed, idx) do
        %{state | streamed_line_count: state.streamed_line_count + 1}
      else
        state
      end

    # Publish text end via EventBus for UI subscribers
    if MapSet.member?(state.banner_printed, idx) do
      EventBus.broadcast_text(
        state.run_id || "",
        state.session_id,
        "\n",
        chunk: true,
        store: false
      )
    end

    cleanup_part(state, idx)
  end

  defp process_event(%Event.ThinkingStart{index: idx} = event, state) do
    fire_callback(
      "part_start",
      %{
        "index" => idx,
        "part_type" => "ThinkingPart",
        "part" => event
      },
      state
    )

    state
    |> add_to_set(:streaming_parts, idx)
    |> add_to_set(:thinking_parts, idx)
    |> emit_thinking_event()
  end

  defp process_event(%Event.ThinkingDelta{index: idx, text: text} = _event, state) do
    fire_callback(
      "part_delta",
      %{
        "index" => idx,
        "delta_type" => "ThinkingPartDelta",
        "delta" => %{content_delta: text}
      },
      state
    )

    if MapSet.member?(state.thinking_parts, idx) do
      emit_thinking_delta_event(state, text)
    else
      state
    end
  end

  defp process_event(%Event.ThinkingEnd{index: idx} = _event, state) do
    fire_callback(
      "part_end",
      %{
        "index" => idx,
        "next_part_kind" => nil
      },
      state
    )

    # Print newline after thinking (if banner was printed)
    state =
      if MapSet.member?(state.banner_printed, idx) do
        %{state | streamed_line_count: state.streamed_line_count + 1}
      else
        state
      end

    cleanup_part(state, idx)
  end

  defp process_event(%Event.ToolCallStart{index: idx, name: name} = event, state) do
    fire_callback(
      "part_start",
      %{
        "index" => idx,
        "part_type" => "ToolCallPart",
        "tool_name" => name,
        "part" => event
      },
      state
    )

    state
    |> add_to_set(:streaming_parts, idx)
    |> add_to_set(:tool_parts, idx)
    |> add_to_set(:banner_printed, idx)
    |> put_tool_name(idx, name || "")
    |> put_token_count(idx, 0)
    |> emit_tool_call_start_event(name)
  end

  defp process_event(%Event.ToolCallArgsDelta{index: idx, arguments: args} = _event, state) do
    fire_callback(
      "part_delta",
      %{
        "index" => idx,
        "delta_type" => "ToolCallArgsDelta",
        "delta" => %{args_delta: args}
      },
      state
    )

    if MapSet.member?(state.tool_parts, idx) do
      # Estimate tokens from args content: rough heuristic, 4 chars ≈ 1 token
      estimated_tokens = max(1, div(byte_size(args || ""), 4))
      current = Map.get(state.token_count, idx, 0)
      new_count = current + estimated_tokens

      tool_name = Map.get(state.tool_names, idx, "")

      # Publish progress via EventBus
      EventBus.broadcast_status(
        state.run_id || "",
        state.session_id,
        "tool_calling",
        metadata: %{
          tool_name: tool_name,
          token_count: new_count,
          index: idx
        }
      )

      %{state | token_count: Map.put(state.token_count, idx, new_count)}
    else
      state
    end
  end

  defp process_event(%Event.ToolCallEnd{index: idx, name: name} = event, state) do
    fire_callback(
      "part_end",
      %{
        "index" => idx,
        "next_part_kind" => nil,
        "tool_name" => name
      },
      state
    )

    # Publish tool call completion via EventBus
    EventBus.broadcast_tool_result(
      state.run_id || "",
      state.session_id,
      name || "",
      %{"completed" => true, "index" => idx},
      tool_call_id: event.id
    )

    cleanup_part(state, idx)
  end

  defp process_event(%Event.UsageUpdate{} = usage, state) do
    fire_callback(
      "usage_update",
      %{
        "prompt_tokens" => usage.prompt_tokens,
        "completion_tokens" => usage.completion_tokens,
        "total_tokens" => usage.total_tokens
      },
      state
    )

    state
  end

  defp process_event(%Event.Done{} = done, state) do
    state =
      fire_callback(
        "done",
        %{
          "id" => done.id,
          "model" => done.model,
          "finish_reason" => done.finish_reason
        },
        state
      )

    # Flush any remaining batched events
    flush_pending_events(state)
  end

  defp process_event(_unknown, state) do
    state
  end

  # ── Part Cleanup ───────────────────────────────────────────────────

  defp cleanup_part(state, idx) do
    state
    |> remove_from_set(:streaming_parts, idx)
    |> remove_from_set(:thinking_parts, idx)
    |> remove_from_set(:text_parts, idx)
    |> remove_from_set(:tool_parts, idx)
    |> remove_from_set(:banner_printed, idx)
    |> then(fn s -> %{s | token_count: Map.delete(s.token_count, idx)} end)
    |> then(fn s -> %{s | tool_names: Map.delete(s.tool_names, idx)} end)
    |> then(fn s -> %{s | text_buffers: Map.delete(s.text_buffers, idx)} end)
  end

  # ── Set Helpers ─────────────────────────────────────────────────────

  defp add_to_set(state, field, idx) do
    Map.update!(state, field, &MapSet.put(&1, idx))
  end

  defp remove_from_set(state, field, idx) do
    Map.update!(state, field, &MapSet.delete(&1, idx))
  end

  # ── Text Buffering ──────────────────────────────────────────────────

  defp put_text_buffer(state, idx, buf) do
    %{state | text_buffers: Map.put(state.text_buffers, idx, buf)}
  end

  defp append_text_buffer(state, idx, text) do
    buffers = Map.update(state.text_buffers, idx, [text], &(&1 ++ [text]))
    %{state | text_buffers: buffers}
  end

  defp maybe_flush_text_buffer(state, idx) do
    case Map.get(state.text_buffers, idx) do
      nil ->
        state

      chunks ->
        buf = Enum.join(chunks)

        if String.contains?(buf, "\n") or byte_size(buf) > @text_flush_char_threshold do
          # Publish the buffered content via EventBus
          EventBus.broadcast_text(
            state.run_id || "",
            state.session_id,
            buf,
            chunk: true,
            store: false
          )

          # Count newlines
          line_count = state.streamed_line_count + count_newlines(buf)

          # Reset buffer
          %{
            state
            | streamed_line_count: line_count,
              text_buffers: Map.put(state.text_buffers, idx, [])
          }
        else
          state
        end
    end
  end

  defp flush_remaining_text_buffer(state, idx) do
    case Map.get(state.text_buffers, idx) do
      nil ->
        state

      [] ->
        state

      chunks ->
        buf = Enum.join(chunks)

        if buf != "" do
          EventBus.broadcast_text(
            state.run_id || "",
            state.session_id,
            buf,
            chunk: true,
            store: false
          )

          line_count = state.streamed_line_count + count_newlines(buf)

          %{
            state
            | streamed_line_count: line_count,
              text_buffers: Map.put(state.text_buffers, idx, [])
          }
        else
          state
        end
    end
  end

  defp count_newlines(str) do
    String.graphemes(str)
    |> Enum.count(&(&1 == "\n"))
  end

  # ── Banner Emission ────────────────────────────────────────────────

  defp maybe_emit_response_banner(state, idx) do
    if MapSet.member?(state.banner_printed, idx) do
      state
    else
      # First content for this text part — emit the response banner
      EventBus.broadcast_status(
        state.run_id || "",
        state.session_id,
        "agent_response",
        metadata: %{banner: true}
      )

      state
      |> add_to_set(:banner_printed, idx)
      |> Map.put(:did_stream_text, true)
      |> Map.put(:streamed_line_count, state.streamed_line_count + 2)
    end
  end

  # TextStart might have initial content — but we defer banner to first delta
  defp maybe_emit_thinking_banner(state, _event, _part_type), do: state

  defp emit_thinking_event(state) do
    # Publish thinking status via EventBus
    EventBus.broadcast_thinking(
      state.run_id || "",
      state.session_id,
      ""
    )

    state
  end

  defp emit_thinking_delta_event(state, text) do
    EventBus.broadcast_thinking(
      state.run_id || "",
      state.session_id,
      text
    )

    state
  end

  defp emit_tool_call_start_event(state, name) do
    EventBus.broadcast_tool_call(
      state.run_id || "",
      state.session_id,
      name || "",
      %{},
      tool_call_id: nil
    )

    state
  end

  # ── Tool Tracking ──────────────────────────────────────────────────

  defp put_tool_name(state, idx, name) do
    %{state | tool_names: Map.put(state.tool_names, idx, name)}
  end

  defp put_token_count(state, idx, count) do
    %{state | token_count: Map.put(state.token_count, idx, count)}
  end

  # ── Callback Firing ────────────────────────────────────────────────

  defp fire_callback(event_type, event_data, state) do
    # Normalize event data for consistent processing by downstream consumers
    normalized_data = EventNormalizer.normalize(event_type, event_data)
    session_id = state.session_id

    entry = {event_type, normalized_data, session_id}
    pending = state.pending_events ++ [entry]

    state = %{state | pending_events: pending}

    # Flush on part_end or when batch threshold reached
    if event_type == "part_end" or length(pending) >= @stream_flush_interval do
      flush_pending_events(state)
    else
      state
    end
  end

  defp flush_pending_events(%{pending_events: []} = state), do: state

  defp flush_pending_events(state) do
    batch = state.pending_events
    state = %{state | pending_events: []}

    # Fire callbacks asynchronously — don't block the handler
    spawn(fn ->
      Enum.each(batch, fn {event_type, event_data, session_id} ->
        try do
          Callbacks.trigger(:stream_event, [event_type, event_data, session_id])
        rescue
          e ->
            Logger.debug("Error flushing stream event: #{inspect(e)}")
        end
      end)
    end)

    state
  end
end
